TELCODOCS-1481- CNF-9121: Reduce DU Profile Apply Time (#76515)
openshift/openshift-docs: modules/ztp-sno-accelerated-ztp.adoc

:_mod-docs-content-type: CONCEPT
[id="ztp-sno-accelerated-ztp_{context}"]
= Accelerated provisioning of {ztp}

:FeatureName: Accelerated provisioning of {ztp}
include::snippets/technology-preview.adoc[]

You can reduce the time taken for installation by using accelerated provisioning of {ztp} for {sno}. 
Accelerated ZTP speeds up installation by applying Day 2 manifests derived from policies at an earlier stage. 

== Activating accelerated ZTP

You can activate accelerated ZTP using the `spec.clusters.clusterLabels.accelerated-ztp` label, as in the following example:

.Example Accelerated ZTP `SiteConfig` CR.
[source,yaml]
----
apiVersion: ran.openshift.io/v2
kind: SiteConfig
metadata:
  name: "example-sno"
  namespace: "example-sno"
spec:
  baseDomain: "example.com"
  pullSecretRef:
    name: "assisted-deployment-pull-secret"
  clusterImageSetNameRef: "openshift-4.10"
  sshPublicKey: "ssh-rsa AAAA..."
  clusters:
  # ...
    clusterLabels:
        common: true
        group-du-sno: ""
        sites : "example-sno"
        accelerated-ztp: full

----

You can use `accelerated-ztp: full` to fully automate the accelerated process. 

> {ztp} updates the `AgentClusterInstall` resource with a reference to the accelerated {ztp} `ConfigMap`, and includes policies extracted from {cgu-operator}, and acclerated ZTP job manifests. 

Reviewer: Includes resources extracted from policies by {cgu-operator}?

Author: ok

> If you use `accelerated-ztp: partial`, {ztp} does not include the accelerated job manifests, but includes policy-derived objects created during the cluster installation, as follows:

Reviewer: includes objects of following kinds? Because the list below specifies the kinds of the objects, not the objects. Just trying to be precise...

Author: precise is good :-)

* `PerformanceProfile.performance.openshift.io`
* `Tuned.tuned.openshift.io`
* `Namespace`
* `CatalogSource.operators.coreos.com`
* `ContainerRuntimeConfig.machineconfiguration.openshift.io`

> This partial acceleration reduces the number of reboots done by the node when applying resources of the kind `Performance Profile`, `Tuned` and `ContainerRuntimeConfig`. 

Reviewer: ```suggestion
This partial acceleration may reduce the number of reboots done by the node when applying resources of the kind `Performance Profile`, `Tuned` and `ContainerRuntimeConfig`. 
```
That's because it depends on many other moving parts. For example, the procedure only works when doing "bootstrap-in-place", but if someone installs OCP using another installation option - it will still reboot all the same. OCP can be installed many different ways...

Author: ok. updated as discussed and added note to mention "Assisted Installer" requirement.

// Module included in the following assemblies:
//
// * edge_computing/ztp-deploying-far-edge-sites.adoc

:_mod-docs-content-type: CONCEPT
[id="ztp-sno-accelerated-ztp_{context}"]
= Accelerated provisioning of {ztp}

:FeatureName: Accelerated provisioning of {ztp}
include::snippets/technology-preview.adoc[]

> You can reduce the time taken for installation by using accelerated provisioning of {ztp} for {sno}. 

Reviewer 2: Suggestion to add "cluster" here for extra clarity, but up to you:
```suggestion
You can reduce the time taken for cluster installation by using accelerated provisioning of {ztp} for {sno}. 
```

Author: thanks for the review @skopacz1 

Yes, makes sense.

Accelerated ZTP speeds up installation by applying Day 2 manifests derived from policies at an earlier stage. 

[IMPORTANT]
====
Accelerated provisioning of {ztp} is supported only when installing {sno} with Assisted Installer. Otherwise this installation method will fail.
====

> == Activating accelerated ZTP

Reviewer 2: Each heading needs to have a unique ID per our [docs guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#anchoring-titles-and-section-headings):
```suggestion
[id="activating-accelerated-ztp_{context}"]
== Activating accelerated ZTP
```

Up to you what the actual ID text is, this is just one suggestion

Author: good suggestion. Thanks.

You can activate accelerated ZTP using the `spec.clusters.clusterLabels.accelerated-ztp` label, as in the following example:

.Example Accelerated ZTP `SiteConfig` CR.
[source,yaml]
----
apiVersion: ran.openshift.io/v2
kind: SiteConfig
metadata:
  name: "example-sno"
  namespace: "example-sno"
spec:
  baseDomain: "example.com"
  pullSecretRef:
    name: "assisted-deployment-pull-secret"
  clusterImageSetNameRef: "openshift-4.10"
  sshPublicKey: "ssh-rsa AAAA..."
  clusters:
  # ...
    clusterLabels:
        common: true
        group-du-sno: ""
        sites : "example-sno"
        accelerated-ztp: full

----

You can use `accelerated-ztp: full` to fully automate the accelerated process. 

> {ztp} updates the `AgentClusterInstall` resource with a reference to the accelerated {ztp} `ConfigMap`, and includes resources extracted from policies by {cgu-operator}, and acclerated ZTP job manifests. 

Reviewer 2: typo: 
```suggestion
{ztp} updates the `AgentClusterInstall` resource with a reference to the accelerated {ztp} `ConfigMap`, and includes resources extracted from policies by {cgu-operator}, and accelerated ZTP job manifests. 
```

> If you use `accelerated-ztp: partial`, {ztp} does not include the accelerated job manifests, but includes policy-derived objects of the following kinds, created during the cluster installation:

Reviewer 2: Suggestion to reword, I feel like "of the following kinds" is a little bit less direct and reads a little awkward:
```suggestion
If you use `accelerated-ztp: partial`, {ztp} does not include the accelerated job manifests, but includes the following policy-derived objects, created during the cluster installation:
```
Up to you though

Author: I take your point but in this case we're referring to the `kind` property of the objects so I think it has to stay.

Reviewer 2: Sounds good, that makes sense! Didn't know "kind" was referring to specific terminology, thanks for clarifying

* `PerformanceProfile.performance.openshift.io`
* `Tuned.tuned.openshift.io`
* `Namespace`
* `CatalogSource.operators.coreos.com`
* `ContainerRuntimeConfig.machineconfiguration.openshift.io`

> This partial acceleration can reduce the number of reboots done by the node when applying resources of the kind `Performance Profile`, `Tuned` and `ContainerRuntimeConfig`. 

Reviewer 2: Similar suggestion as above, in this case "of the kind". Would this sentence still mean the same thing if it was reworded like this?
```suggestion
This partial acceleration can reduce the number of reboots done by the node when applying `Performance Profile`, `Tuned` and `ContainerRuntimeConfig` resources. 
```

Once again though I'll leave that up to you

Author: Thanks, same as above

> This partial acceleration can reduce the number of reboots done by the node when applying resources of the kind `Performance Profile`, `Tuned` and `ContainerRuntimeConfig`. 

Reviewer 2: In either case, I think this sentence needs a serial/Oxford comma per the [ISG](https://www.ibm.com/docs/en/ibm-style?topic=punctuation-commas#commas-between-items-in-a-series):
```suggestion
This partial acceleration can reduce the number of reboots done by the node when applying resources of the kind `Performance Profile`, `Tuned`, and `ContainerRuntimeConfig`. 
```

Author: yes. thanks.

> {cgu-operator} installs the operator subscriptions derived from policies after {rh-rhacm} completes the import of  the cluster, following the same flow as standard {ztp}.

Reviewer 2: Operator should be capitalized, and there's an extra space here:
```suggestion
{cgu-operator} installs the Operator subscriptions derived from policies after {rh-rhacm} completes the import of the cluster, following the same flow as standard {ztp}.
```

The benefits of accelerated ZTP increase with the scale of your deployment. 
Using `accelerated-ztp: full` gives more benefit on a large number of clusters. 
With a smaller number of clusters, the reduction in installation time is less significant.

> Full acclerated ZTP leaves behind a namespace and a completed job on the spoke that need to be manually removed.

Reviewer 2: Same typo:
```suggestion
Full accelerated ZTP leaves behind a namespace and a completed job on the spoke that need to be manually removed.
```

One benefit of using `accelerated-ztp: partial` is that you can override the functionality of the on-spoke job if something goes wrong with the stock implementation or if you require a custom functionality.

> == The accelerated ZTP process

Reviewer 2: Same comment as above about including an ID:
```suggestion
[id="the-accelerated-ztp-process_{context}"]
== The accelerated ZTP process
```

Up to you what the actual ID text is, this is just an idea

Accelerated ZTP uses an additional `ConfigMap` to create the resources derived from policies on the spoke cluster.  
The standard `ConfigMap` includes manifests that the {ztp} workflow uses to customize cluster installs. 
{cgu-operator} detects that the `accelerated-ztp` label is set and then creates a second `ConfigMap`.
As part of accelerated ZTP, the `SiteConfig` generator adds a reference to that second `ConfigMap` using the naming convention `<spoke-cluster-name>-aztp`. 
After {cgu-operator} creates that second `ConfigMap`, it finds all policies bound to the managed cluster and extracts the {ztp} profile information. 

> {cgu-operator} adds the {ztp} profile information to the `<spoke-cluster-name>-aztp` `ConfigMap` CR and applies the CR to the hub cluster API. 

Reviewer 2: I'd personally suggest finding a place in here to introduce a paragraph break. The content is fine but I believe that it does slightly look like a wall of text in the rendered output, which can impact its readability

Author: yes, a bit hard on the eyes. I've broken it up for readability.

> {cgu-operator} adds the {ztp} profile information to the `<spoke-cluster-name>-aztp` `ConfigMap` CR and applies the CR to the hub cluster API. 

Reviewer 2: Nit - since this is the first mention of CR in the section, I'd just write it out on this first instance:
```suggestion
{cgu-operator} adds the {ztp} profile information to the `<spoke-cluster-name>-aztp` `ConfigMap` custom resource (CR) and applies the CR to the hub cluster API. 
```

Author: ok
